### PR TITLE
feat: adding slim dockerfile

### PR DESCRIPTION
**Pull request by johnson2427**
### What I did

Created Dockerfile.slim for a slim version of ape without recommended-plugins

fixes: #2143 

### How I did it

Needed to create the whl files with python:3.11, then switch to python:3.11-slim after the whl files are created. This install ends up being 500MB in size vs 2.24GB in the current Dockerfile build.

I then install the recommended plugins in the Dockerfile for the latest version. The size of the image reduced from 2.24GB to 571MB.

### How to verify it

```bash
git checkout feat/create-bare-ape-image
docker build -t ape:latest-slim -f Dockerfile.slim .
docker image ls
...
...      ...               ...   SIZE
ape   latest-slim                500MB
ape   latest                     571MB
```

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: Dockerfile.slim
Comment (antazoey) [inline Dockerfile.slim:22]: Question: Would there be a nice way to share Dockerfile-logic between this one and the other one? I noticed they do some of the same things, and I don't want to update one and forget to update the other one.

**Review by antazoey — APPROVED** [review]
One question, with an idea for potential improvement, but otherwise LGTM

File: Dockerfile.slim
Comment (fubuloubu) [reply Dockerfile.slim:22]: I think as long as `-slim` was built first, that this one could use `-slim` as it's base to install `recommended-plugins` over

Comment (johnson2427) [reply Dockerfile.slim:22]: @fubuloubu you are correct, I'm just working on the implementation locally to ensure I'm not missing anything

Comment (johnson2427) [reply Dockerfile.slim:22]: I'm having issues keeping this in the python:3.11-slim though, which is the hold up

**Comment by johnson2427** [comment]
@fubuloubu make sure this works the way it did before. I checked the pip list, and for some reason this is installing eth-ape in two places, so I'm still trying to figure that out.

File: Dockerfile.slim
Comment (antazoey) [inline Dockerfile.slim:30]: this may need to be a parameter, as it will be the same otherwise for both slim and fat

Comment (johnson2427) [reply Dockerfile.slim:30]: I think it should be the same for both

Comment (antazoey) [reply Dockerfile.slim]: Ya that makes sense. Alternatively, you can exclude from the slim

File: Dockerfile.slim
Comment (fubuloubu) [inline Dockerfile.slim:27]: Wonder if OCI is better:
https://github.com/opencontainers/image-spec/blob/main/annotations.md#pre-defined-annotation-keys

https://snyk.io/blog/how-and-when-to-use-docker-labels-oci-container-annotations/

File: Dockerfile.slim
Comment (fubuloubu) [inline Dockerfile.slim:35]: This is just a general guide? Think this can be shortened
```suggestion
      org.label-schema.docker.cmd="docker run --volume $PWD:/home/harambe/project --workdir /home/harambe/project apeworx/ape run my_script.py"
```

Or something like that since we are not bundling any compilers here

File: Dockerfile.slim
Comment (fubuloubu) [inline Dockerfile.slim:43]:
```suggestion
```

File: Dockerfile
Comment (fubuloubu) [inline Dockerfile:1]: This one should keep the stuff particular to `recommended-plugins`

**Review by fubuloubu — COMMENTED** [review]
Keep coming back to our [github-action](https://github.com/ApeWorX/github-action) and wondering if there is a good way to make the `recommended-plugins` image work with the action to speed that up

Sorry about all the comments, didn't realize just how out of date the old dockerfile was

File: .github/workflows/build.yaml
Comment (antazoey) [reply .github/workflows/build.yaml:32]: There is a `fetch-tags` parameter on the checkout action we could use.
Tags are needed for setuptools-scm to install the correct version.

File: Dockerfile.slim
Comment (johnson2427) [reply Dockerfile.slim:43]: I'll try to find a way to do this outside of slim, I'm currently creating the wheels for recommended plugins, but not installing them.

We install them in the Dockerfile. It isn't ideal, I'll fix this

File: .github/workflows/build.yaml
Comment (johnson2427) [reply .github/workflows/build.yaml:36]: absolutely, will add when I get the build and push to work

File: .github/workflows/build.yaml
Comment (johnson2427) [reply .github/workflows/build.yaml:37]: we're going to create 2 tags here. One will be latest-slim and then a second for the version release (0.8.13-slim). That way, you'll reach out to latest for this release, when the next release comes out, the `latest-slim` tag will be replaced, but the 0.8.13-slim will remain in place.

File: Dockerfile.slim
Comment (johnson2427) [reply Dockerfile.slim:22]: This has been updated. Slim builds with `Dockerfile.slim`, recommended plugins is built in the `Dockerfile`. The logic is tidied up here now.

File: .github/workflows/build.yaml
Comment (johnson2427) [reply .github/workflows/build.yaml:31]: added this, checking that it works

Comment (johnson2427) [reply .github/workflows/build.yaml]: This worked!

File: .github/workflows/build.yaml
Comment (fubuloubu) [reply .github/workflows/build.yaml:36]: oh, usually I have:
- `latest` = whatever is head on `main` branch
- `stable` = last release
- `vX.Y.Z` = specific release

and then to that we do `{tag}-slim` for the slim build

File: .github/workflows/build.yaml
Comment (johnson2427) [reply .github/workflows/build.yaml:1]: @fubuloubu I **think** this is what we want to do. Letting this build now. The only issue is, I'll need the fat image to pull the slim image. I have to look at letting it reference the build rather than pulling from ghcr

File: .github/workflows/build.yaml
Comment (fubuloubu) [inline .github/workflows/build.yaml:39]: Add a debug step to see the tags and labels

File: .github/workflows/build.yaml
Comment (fubuloubu) [inline .github/workflows/build.yaml:52]: Forgot this?
```suggestion
          file: ./Dockerfile.slim
          push: ${{ github.event_name != 'pull_request' }}
```

File: .github/workflows/build.yaml
Comment (fubuloubu) [inline .github/workflows/build.yaml:53]: Does this produce two tags when published? (`v0.X.y` and `stable`)

File: .github/workflows/build.yaml
Comment (johnson2427) [reply .github/workflows/build.yaml:1]: There is no way to use the docker engine with cache in github workflows. So I have to push the slim. I don't have to push a stable. Slim will push with a pr-xxxx label

File: .github/workflows/build.yaml
Comment (johnson2427) [reply .github/workflows/build.yaml:58]: I can't build the fat image without the slim being built with our current setup. If we want to do that, the Dockerfile will have to duplicate the code from Dockerfile.slim

File: .github/workflows/build.yaml
Comment (johnson2427) [reply .github/workflows/build.yaml:59]: Hard to know at the moment. What I did should work, but I'm not certain until we push.

File: Dockerfile.slim
Comment (johnson2427) [reply Dockerfile.slim:21]: Let me know what you want to do here

Comment (antazoey) [reply Dockerfile.slim:21]: Yes, OCI is more standard now. Also http://label-schema.org is does not work on https which is sus

Comment (fubuloubu) [reply Dockerfile.slim:21]: It also just says it is deprecated and use OCI

Comment (johnson2427) [reply Dockerfile.slim:21]: OCI added

**Comment by johnson2427** [comment]
This is what we end up with
```bash
            "build-arg:BUILD_DATE": "2024-08-16T18:48:15Z",
            "build-arg:VCS_REF": "c3eada713fcf78b7fba8d82571298869883ffb14",
            "build-arg:VERSION": "",
            "label:org.opencontainers.image.created": "2024-08-16T19:13:35.036Z",
            "label:org.opencontainers.image.description": "The smart contract development tool for Pythonistas, Data Scientists, and Security Professionals",
            "label:org.opencontainers.image.licenses": "Apache-2.0",
            "label:org.opencontainers.image.revision": "c3eada713fcf78b7fba8d82571298869883ffb14",
            "label:org.opencontainers.image.source": "https://github.com/ApeWorX/ape",
            "label:org.opencontainers.image.title": "ape",
            "label:org.opencontainers.image.url": "https://github.com/ApeWorX/ape",
            "label:org.opencontainers.image.version": "pr-2219"
```

**Comment by johnson2427** [comment]
I need to fix the documentation for the `Intalling with 'docker'` section, but I'm having a hard time doing so without publishing the images. I think we should update docs in a separate PR, but I'll leave that up to you two

**Comment by fubuloubu** [comment]
> I need to fix the documentation for the `Intalling with 'docker'` section, but I'm having a hard time doing so without publishing the images. I think we should update docs in a separate PR, but I'll leave that up to you two

How do you think it will be different?

**Comment by johnson2427** [comment]
> > I need to fix the documentation for the `Intalling with 'docker'` section, but I'm having a hard time doing so without publishing the images. I think we should update docs in a separate PR, but I'll leave that up to you two
> 
> How do you think it will be different?

We are no longer going to be pushing to dockerhub from what I remember in our conversation from earlier this week. I'd like to give the `ghcr` location for users.

**Comment by fubuloubu** [comment]
> > > I need to fix the documentation for the `Intalling with 'docker'` section, but I'm having a hard time doing so without publishing the images. I think we should update docs in a separate PR, but I'll leave that up to you two
> > 
> > How do you think it will be different?
> 
> We are no longer going to be pushing to dockerhub from what I remember in our conversation from earlier this week. I'd like to give the `ghcr` location for users.

Ah yes, for sure. I think what we can do with this PR is stop publishing to dockerhub and start publishing to ghcr, and then update the instructions before we create a release containing this update. Once merged, I will stop automated publishing on dockerhub, and we should also port some of the older images to ghcr as well to ensure continuity for any users, and then make a public announcement for that as well

**Comment by johnson2427** [comment]
Currently, from the automated metadata creation, this is where our image will go `ghcr.io/apeworx/ape`. I believe you want this to go to `ghcr.io/apeframework/ape`, correct @fubuloubu?

**Comment by fubuloubu** [comment]
> Currently, from the automated metadata creation, this is where our image will go `ghcr.io/apeworx/ape`. I believe you want this to go to `ghcr.io/apeframework/ape`, correct @fubuloubu?

No! That place is good, wanted it @ ghcr.io/apeworx/ape

File: README.md
Comment (fubuloubu) [inline README.md:62]:
```suggestion
You can pull our images from [ghcr](https://ghcr.io/apeworx/ape).
We currently have a `slim` docker image that is built without any installed plugins, as well as an image that is built with all plugins included from our `recommended-plugins` extra.
```

File: README.md
Comment (fubuloubu) [inline README.md:70]: I think put the other image first, and not recommended to install w/ `slim` unless you need to

Comment (johnson2427) [reply README.md]: lmk if that's what you're looking for

File: README.md
Comment (fubuloubu) [inline README.md:63]:
```suggestion
This image is built using our `recommended-plugins` extra, so it is a great starting point for running ape in a containerized environment.

We also have a `slim` docker image that is built without any installed plugins.
This image is meant for production support and must be further configured if any plugins are in use.
```

File: README.md
Comment (fubuloubu) [inline README.md:96]:
```suggestion
> The above command requires the full install whicy includes `recommended-plugins` installation extra.
```

**Review by fubuloubu — APPROVED** [review]
If you feel confident in the workflow, it is ready to go. Would recommend just doing some local testing with a repo under your personal if you've not sure

**Pull request merged**
